Updates of checking file size of firstguess and a few more

GangZhao-NOAA opened the pull request:
### Some updates to hrrr_3drtma:
   
1. In prepfgs script, when searching for the firstguess file in HRRR forecast files,  check the file size to make sure the firstguess file is good (simple data integrity check);
2. Making the updatevars step running under its own working directory (wrfprd), instead of under gsiprd;
3. Set up a share sub-directory under each cycle running top directory, and move or copy the data which need to be shared to other steps.  For example, wrf_inout, pe*.nc4 files generated by gsianl step, the wrf_inout would be used in updatevars and anldump steps, and the split diag files pe*nc4 would be used by ncdiag to combine them.
4. In util_dev/getbest_EnKF_FV3GDAS.py, it links the gdas ensemble to local running directoy. This linking action is not necessary if gdas ensemble would not be used in analysis. So the linking lines are commented off in util_dev/getbest_EnKF_FV3GDAS.py. 
5. Correspondingly, in gsianl script, if the gdas ensemble is used in hybrid analysis, then the gdas ensemble members are links to local analysis working directory.
6. In gsianl step, at the end of script, Using cfp command to parallelly copy the thousand of split netcdf-format obs-diag files to shared sub-directory. It might take 6-10 mins if copying these files serially.
7. A few minor issues:, such as in wrfare building script, adding a line to create a sub-directory "run/" in order to avoid the annoying warning message at the end of building.

GangZhao-NOAA committed:
Some updates to hrrr_3drtma:
1. udpatevars step runs under its own working directory (wrfprd), instead of under gsiprd;
2. set up a share sub-directory under each cycle running top directory, and move or copy
   the data which need to be shared to other steps.
   For example, wrf_inout, pe*.nc4 files generated by gsianl step,
   the wrf_inout would be used in updatevars and anldump steps, and
   the split diag files pe*nc4 would be used by ncdiag to combine them.
3. In util_dev/getbest_EnKF_FV3GDAS.py, it links the gdas ensemble to local running directoy.
   This linking action is not necessary if gdas ensemble would not be used in analysis. So
   the linking lines are commented off in util_dev/getbest_EnKF_FV3GDAS.py.
   Correspondingly, in gsianl script, if the gdas ensemble is used in hybrid analysis,
   then the gdas ensemble members are links to local analysis working directory.
4. A few minor issues:, such as in wrfare building script, adding a line to create a
   sub-directory "run/" in order to avoid the annoying warning message at the nd of building.

GangZhao-NOAA committed:
A few more updates to hrrr_3drtma:
1. In prepfgs script, When searching for the firstguess file in HRRR forecast files,
   check the file size to make sure the firstguess file is good (simple data integrity check);
2. In gsianl step, at the end of script, Using cfp command to parallelly copy the
   distributed netcdf-format obs-diag files to shared sub-directory.
   (It might take 6-10 mins if copy these files serially.)

AnnetteGibbs-NOAA commented on jobs/JRTMA3D_CLEAN:
```diff
@@ -45,6 +45,19 @@ export HOMErtma3d=${HOMErtma3d:-$NWROOT/rtma.${rtma_ver}}
 # date and time for previous and next cycle
 export PDYHH="${PDY}${cyc}"
 
+# creating a sub-directory under top running directory to share the data among various steps
```
Hi Gang, I am curious why you chose the "clean" job script to create your shared directory.  I would think you would want to create your shared directory in the GSI script and it looks like that is what you did.  So I am not sure if these changes to JRTMA3D_CLEAN are needed, when I think you can do them all in the GSI script.

GangZhao-NOAA replied:
Hi @AnnetteGibbs-NOAA,

Thank you for pointing this out. 

Yes, in current run, the "shared" directory is only needed after the GSI analysis, so I did create this shared directory in GSI script. Later I thought if this shared directory might be necessary for other steps, such as the prepfgs or prepobs, then it might be better to create the shared directory in the beginning of the cycle. "Clean" step seems to be the first step of each cycle in the workflow (in the xml file), so I also added the lines to create shared directory in Clean script.

Thanks,
-Gang

AnnetteGibbs-NOAA commented on scripts/rtma3d/exrtma3d_gsianl.ksh:
```diff
@@ -254,6 +254,18 @@ elif [[ ${nummem} -eq 80 ]]; then
      HYBENS_INFO=${PARMgsi}/hybens_info
      ${CP} ${PARMgsi}/hybens_info hybens_info
   fi
+  # link GDAS ensemble members to the analysis working directory
```
Were you able to test this section of the code to make sure it would work in case we need to use the GDAS ensemble on the CONUS grid?  If not, I can test it when I resume the NCO compliance work.

GangZhao-NOAA replied:
Hi @AnnetteGibbs-NOAA 

Ah, that's a good point. To be honest, I have not tested it yet, in my testrun I only run wth HRRRDAS ensemble. It's my bad, I will test it in both CONUS and ALASKA parallels.

Thank you!
-Gang

AnnetteGibbs-NOAA replied:
If you have to move onto something else, I can just test it when I resume the compliance work.

MatthewMorris-NOAA commented on jobs/JRTMA3D_UPDATEVARS:
```diff
@@ -129,18 +129,22 @@ if [ "${envir}" != "esrl" ]; then #WCOSS
+#    wrfprd : the updatevars working directory (model 1-step forecast) 
+    export DATAHOME=${DATA_WRFPRD:-"${DATA_RUNDIR}/wrfprd"}
+    if [ -L "${DATAHOME}" ] && [ -d "${DATAHOME}" ] ; then
+        ${RM} -rf ${DATAHOME}
```
It may be better to use "unlink" here.

GangZhao-NOAA replied:
Hi @MatthewMorris-NOAA 
Yes, "unlink" is a good option here. Since the current command "rm -rf " is still working here, I think it is fine to keep using it. I noticed that there are quite a few similar lines in j-job scripts, we may replace them with "unlink" in the later PR.
Thanks,
-Gang

MatthewMorris-NOAA replied:
Hi @GangZhao-NOAA OK, that sounds good, thanks!

AnnetteGibbs-NOAA commented on scripts/rtma3d/exrtma3d_prepfgs.ksh:
```diff
@@ -102,19 +107,49 @@ CDATE=$PDY$cyc
+         if [[ ${size_match} =~ [yYtT] ]] ; then
+             cpreq $GESINhrrr/${probe_hrrr_guess_nc} ${DATA}/${FGSrtma3d_FNAME}
+             break      # breaking out the loop of searching for firstguess in HRRR forecast
+         else
+             msg="hrrr ${ic}-hour forecast file ${probe_hrrr_guess_nc} exists, but its filesize (${filesize}) does not match the standard size (${targetsize_hrrr}) even after waiting for ${ics_max} minutes. Trying to search in the earlier hrrr forecast files ..."
```
Did you test the option where the filesize did not match the standard size after waiting 15 minutes?  If not, I will test with the NCO compliance work.

GangZhao-NOAA replied:
Hi @AnnetteGibbs-NOAA ,

You got me. I have not tested these lines in real run, because there are very rare case that the file is incomplete. I need to make a fake incomplete file and put it under another directory to test (since I could not put it under hrrr prod directory). 
This is a good question. I do need to test it. I will make a fake one to test it. I will let you know how it goes.

Thanks
-Gang

AnnetteGibbs-NOAA replied:
Or I can just test it when I do the compliance work.  Either way works for me.  Thanks Gang!

GangZhao-NOAA replied:
@AnnetteGibbs-NOAA 
I am building up the parallel on WCOSS2 now. If I can make the test with the fake file soon, I will let you know.
Thank you!
-Gang

MatthewMorris-NOAA commented on scripts/rtma3d/exrtma3d_anldump.ksh:
```diff
@@ -24,11 +24,14 @@ check_dirs_exist() { #usage: check_dirs_exist "var1_name" "var2_name" ...
 
-  ncf_anl=${COMOUTgsi_rtma3d}/${ANLrtma3d_FNAME}
+# ncf_anl=${COMOUTgsi_rtma3d}/${ANLrtma3d_FNAME}
+  ncf_anl=${DATA_SHARED}/wrf_inout
```
Why is the shared directory needed in the "anldump" step?

GangZhao-NOAA replied:
Hi @MatthewMorris-NOAA 
anl_dump step is going to dump out the analyses of "HOWV", "GUST' and "VIS" from the netcdf format GSI analysis file --"wrf_inout, then re-write them in grib2 files for the follow-up post-processing or smartinit steps. Since the analysis file "wrf_inout" has been copied to shared directory, so anl_dump step also needs to know about the "shared" directory.

Thanks,
-Gang

MatthewMorris-NOAA replied:
OK, got it!

AnnetteGibbs-NOAA reviewed (approved):
Thanks for answering all my questions.  Looks good Gang.  I approve your PR.

MatthewMorris-NOAA commented on scripts/akrtma3d/exakrtma3d_gsianl.ksh:
```diff
@@ -192,7 +192,7 @@ nummem=`more filelist03 | wc -l`
 hrrrmem=`more filelist.hrrrdas | wc -l`
 hrrrmem=$((hrrrmem - 3 ))
-if [[ ${hrrrmem} -gt 30 ]] && [[ ${HRRRDAS_BEC} -eq 1  ]]; then #if HRRRDAS BEC is available, use it as first choice
+if [[ ${hrrrmem} -gt 30 ]] && [[ ${HRRRDAS_BEC} -ge 1  ]]; then #if HRRRDAS BEC is available, use it as first choice
```
This is just a general comment, not related to this PR. Why is this seciton included here if there's no HRRRDAS for Alaska?

GangZhao-NOAA replied:
Hi @MatthewMorris-NOAA 
You are right. This part is actually not necessary for current Alaska run. We can remove it in the later PR.
-Gang

MatthewMorris-NOAA commented:
Hi @GangZhao-NOAA I've gone through the changes, and everything generally looks good.  Do you want me to wait for you to finish testing the firstguess filesize check before approving the PR?

Also, I'm wondering if we should include a similar check for Alaska.  While there haven't been recent failures on that domain, I recall there were some segmentation fault failures several months ago.  Looking at the HRRR production scripts, it seems they do include a check for Alaska. /lfs/h1/ops/prod/packages/hrrr.v4.1.20/scripts/alaska/exhrrr_analysis.sh

I'd also be fine with merging this PR and adding the filesize check for Alaska later on.

GangZhao-NOAA commented:
> Hi @GangZhao-NOAA I've gone through the changes, and everything generally looks good. Do you want me to wait for you to finish testing the firstguess filesize check before approving the PR?
> 
> Also, I'm wondering if we should include a similar check for Alaska. While there haven't been recent failures on that domain, I recall there were some segmentation fault failures several months ago. Looking at the HRRR production scripts, it seems they do include a check for Alaska. /lfs/h1/ops/prod/packages/hrrr.v4.1.20/scripts/alaska/exhrrr_analysis.sh
> 
> I'd also be fine with merging this PR and adding the filesize check for Alaska later on.

Hi @AnnetteGibbs-NOAA @MatthewMorris-NOAA 
If you do not have Pull Request recently (this week), we can hold this PR until I finish the test successfully for CONUS and Alaska and for fake imcomplete firstguess file case. 
Thank you!
-Gang

MatthewMorris-NOAA commented:
Hi @GangZhao-NOAA, OK, I will hold off on approving the PR until you can test the changes. Thanks!

GangZhao-NOAA committed:
further modifications for PR #81
1. adding firstguess file size check for Alaska domain (in scripts/akrtma3d/exakrtma3d_prepfgs.ksh)
2. fixed bug (failed to create working directory wrfprd.$jobid} and its symbolic link wrfprd)
   (define rundir_task in workflow/rtma3d_pbspro_conus.xml, workflow/rtma3d_pbspro_alaska.xml for
     updatevars_task, and create $DATA directory in jobs/JRTMA3D_UPDATEVARS)
3. increasing wallclock time limit of step prepfgs to 30 mins (since it might need to wait 15 mins
     for file size check);

GangZhao-NOAA commented:
@MatthewMorris-NOAA @AnnetteGibbs-NOAA 
In the real-time parallel test on WCOSS2, I found a few bugs and fixed them. The updates were in the [commit #90bcdd](https://github.com/GangZhao-NOAA/NOAA_3drtma/commit/90bcdd840e53f45f49585c7baecfc3ff1affdf95) .
Now the modified code was tested successfully, especially the file-size check also was checked and it works well for both CONUS and Alaska.
You may find the log files on WCOSS2 (Dev)
under "/lfs/h2/emc/ptmp/gang.zhao/rtma3d_logs"
Especially for file size check, I tested it for cycle  20251010_12:00Z, so 
for CONUS:
/lfs/h2/emc/ptmp/gang.zhao/rtma3d_logs/conus/:
rtma3d_pbspro_prepfgs_2025101012.log:   it found the filesize error for a faked file  (1-hr forecast)"/lfs/h2/emc/ptmp/gang.zhao/Cases_HRRR_3DRTMA/hrrr_fgs/hrrr/v4.1/nwges/hrrrges_sfc/conus/hrrr_2025101011f001", then after trying 5 times, then using 2-hr forecast
"/lfs/h2/emc/ptmp/gang.zhao/Cases_HRRR_3DRTMA/hrrr_fgs/hrrr/v4.1/nwges/hrrrges_sfc/conus/hrrr_2025101010f002"

For Alaska:
under /lfs/h2/emc/ptmp/gang.zhao/rtma3d_logs/alaska/:
akrtma3d_pbspro_prepfgs_ind3_2025101012.log: it found the filesize error for a faked file (3-hr forecast) "/lfs/h2/emc/ptmp/gang.zhao/Cases_HRRR_3DRTMA/hrrr_fgs/hrrr/v4.1/nwges/hrrrges_sfc/alaska/hrrrak_2025101009f003", after trying 5 times, then using 6-hr forecast
"/lfs/h2/emc/ptmp/gang.zhao/Cases_HRRR_3DRTMA/hrrr_fgs/hrrr/v4.1/nwges/hrrrges_sfc/alaska/hrrrak_2025101006f006".

Please review the updated PR #81.

Thank you!

-Gang

MatthewMorris-NOAA reviewed (approved):
Looks good @GangZhao-NOAA, approved!

GangZhao-NOAA committed:
adding mail functionality when file check fails for firstguess (PR #81)
  if firstguess file is missing, or filesize does not match, warning message
  is sent out to user by email.

GangZhao-NOAA commented:
Hi @MatthewMorris-NOAA @AnnetteGibbs-NOAA @ManuelPondeca-NOAA,
I finally added the email functionality with filesize check in prepfgs step (for noth CONUS and Alaska). Please review the updated PR #81 . You may only review the 3 updated files through [this link](https://github.com/NOAA-EMC/NOAA_3drtma/pull/81/commits/f6578c305bf2ae650a484070988b16e3fe94f4fc)
Sorry for sending out the review request during the weekend. There is no rush. You can do it next Monday.
Thank you! Have a good weekend!
-Gang

AnnetteGibbs-NOAA reviewed (approved):
Looks good, Gang.  I approve these changes.

GangZhao-NOAA commented:
@AnnetteGibbs-NOAA @MatthewMorris-NOAA
Thank you very much for the review!
Now it is merged.
